crates/runtime/src/debugger.rs

```diff
@@ -0,0 +1,163 @@
+//! Debugger utils.
+#![allow(missing_docs)]
+use std::any::Any;
+
+/// Patching the code trait
+pub trait PatchableCode {
+    fn patch_jit_code(&mut self, addr: usize, len: usize, f: &mut dyn FnMut());
+}
+
+pub struct BreakpointData {
+    /// Breakpoint address.
+    pub pc: *const u8,
+    data: SavedBreakpointData,
+}
+
+impl BreakpointData {
+    /// Creates new BreakpointData if possible.
+    pub fn new(pc: *const u8, patchable: &mut dyn PatchableCode) -> Option<Self> {
+        let mut data = None;
+        patchable.patch_jit_code(pc as usize, BREAKPOINT_INSTR_LEN, &mut || {
+            data = Some(unsafe { write_breakpoint_instr(pc as usize) });
+        });
+        data.map(|data| BreakpointData { pc, data })
+    }
+
+    pub fn restore_code(&self, patchable: &mut dyn PatchableCode) {
+        let pc = self.pc as usize;
+        patchable.patch_jit_code(pc, BREAKPOINT_INSTR_LEN, &mut move || unsafe {
+            self.data.restore(pc);
+        });
+    }
+
+    pub fn patch_code(&self, patchable: &mut dyn PatchableCode) {
+        let pc = self.pc as usize;
+        patchable.patch_jit_code(
+            self.pc as usize,
+            BREAKPOINT_INSTR_LEN,
+            &mut move || unsafe {
+                write_breakpoint_instr(pc);
+            },
+        );
+    }
+}
+
+const BREAKPOINT_INSTR: u8 = 0xCC;
+const BREAKPOINT_INSTR_LEN: usize = 1;
```

Comment:
Make this support AArch64?

Reply:
Yes, I need guidance here, I think just cfg() here for x86.

- [ ] Support for AArch64